host/sgxsign.c

```diff
@@ -71,50 +79,238 @@ static oe_result_t _get_date(unsigned int* date)
+static oe_result_t _get_modulus(
+    const oe_rsa_public_key_t* rsa,
+    uint8_t modulus[OE_KEY_SIZE])
 {
     oe_result_t result = OE_UNEXPECTED;
     uint8_t buf[OE_KEY_SIZE];
+    size_t bufsize = sizeof(buf);
 
     if (!rsa || !modulus)
         OE_RAISE(OE_INVALID_PARAMETER);
 
-    if (!BN_bn2bin(rsa->n, buf))
+    OE_CHECK(oe_rsa_public_key_get_modulus(rsa, buf, &bufsize));
+
+    /* RSA key length is the modulus length, so these have to be equal. */
+    if (bufsize != OE_KEY_SIZE)
         OE_RAISE(OE_FAILURE);
 
-    _mem_reverse(modulus, buf, OE_KEY_SIZE);
+    _mem_reverse(modulus, buf, bufsize);
 
     result = OE_OK;
 
 done:
     return result;
 }
 
-static oe_result_t _get_exponent(RSA* rsa, uint8_t exponent[OE_EXPONENT_SIZE])
+static oe_result_t _get_exponent(
+    const oe_rsa_public_key_t* rsa,
+    uint8_t exponent[OE_EXPONENT_SIZE])
 {
     oe_result_t result = OE_UNEXPECTED;
-    // uint8_t buf[OE_EXPONENT_SIZE];
+    uint8_t buf[OE_EXPONENT_SIZE];
+    size_t bufsize = sizeof(buf);
 
     if (!rsa || !exponent)
         OE_RAISE(OE_INVALID_PARAMETER);
 
-    if (rsa->e->top != 1)
-        OE_RAISE(OE_FAILURE);
+    OE_CHECK(oe_rsa_public_key_get_exponent(rsa, buf, &bufsize));
+
+    /* Exponent is in big endian. So, we need to reverse. */
+    _mem_reverse(exponent, buf, bufsize);
+
+    /* We zero out the rest to get the right exponent in little endian. */
+    OE_CHECK(
+        oe_memset_s(
+            exponent + bufsize,
+            OE_EXPONENT_SIZE - bufsize,
+            0,
+            OE_EXPONENT_SIZE - bufsize));
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+#if defined(_WIN32)
+static oe_result_t _calc_q1_q2_bignum(
+    const unsigned char* signature,
+    size_t signature_size,
+    const unsigned char* modulus,
+    size_t modulus_size,
+    mbedtls_mpi* q1,
+    mbedtls_mpi* q2)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    mbedtls_mpi s;
+    mbedtls_mpi m;
+    mbedtls_mpi r1;
+    mbedtls_mpi t1;
+
+    mbedtls_mpi_init(&s);
+    mbedtls_mpi_init(&m);
+    mbedtls_mpi_init(&r1);
+    mbedtls_mpi_init(&t1);
+
+    if (!signature || !signature_size || !modulus || !modulus_size || !q1 ||
+        !q2)
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    /* Create new objects */
+    {
+        if (mbedtls_mpi_read_binary(&s, signature, signature_size) != 0)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+
+        if (mbedtls_mpi_read_binary(&m, modulus, modulus_size) != 0)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+    }
 
+    /*
+     * Intel SGX docs state that Q1 and Q2 should store the following values.
+     *   - Q1 = FLOOR(signature^2 / modulus)
+     *   - Q2 = FLOOR(signature^3 - Q1 * signature * modulus) / modulus
+     *
+     * These values are used to optimize the RSA signature verification,
+     * which is normally calculated as S^3 mod M. We see that we can
+     * derive these Q1 & Q2 values:
+     *  - S^3 mod M
+     *    -> (S * (S^2 mod M)) mod M
```

Comment:
(a * b) mod n = ((a mod n) * (b mod n)) mod n
With a = S, b = S^2, and n = M
shouldn't this be: ((S modM) * (S^2 mod M)) mod M?

Reply:
Yeah, I skipped a step here:

The property is that `(a * b) mod n = (( a mod n) * (b * mod n)) mod n`. To calculate the RSA signature `S`, you do `S = (plaintext)^d mod n` , so we have `S < n` and consequently, `S = (S mod n)`.